src/parser.rs

```diff
@@ -58,53 +63,55 @@ pub struct Tag<'a> {
 }
 
 pub fn tag_header(input: &[u8]) -> IResult<&[u8], TagHeader> {
-  let (i, tag_type) = be_u8(input)?;
-  let tag_type = match tag_type {
-    8 => TagType::Audio,
-    9 => TagType::Video,
-    18 => TagType::Script,
-    _ => return Err(Err::Error(Error::new(input, ErrorKind::Alt))),
-  };
-  let (i, data_size) = be_u24(i)?;
-  let (i, timestamp) = be_u24(i)?;
-  let (i, timestamp_extended) = be_u8(i)?;
-  let (i, stream_id) = be_u24(i)?;
-  let header = TagHeader {
-    tag_type,
-    data_size,
-    timestamp: (u32::from(timestamp_extended) << 24) + timestamp,
-    stream_id,
-  };
-  Ok((i, header))
+  be_u8(input).and_then(|(i, tag_type)| {
```

Comment:
Why not make the whole thing a tuple that includes the initial byte?

Reply:
Since we check `tag_type`, I thought it would have been better to do that before running the other functions

Reply:
The components of the tuple run one after another, and the error of the first one that fails would be returned. So I think this would be equivalent or am I missing something?

Reply:
Yep, but I was talking about this line https://github.com/rust-av/flavors/pull/28/files#diff-4a04259da480a6b794a2e947e4cc03eff4d1aa9330836f5b91cac68c5398193fR71 that is related to the first parser. Or am I missing something too?

Reply:
As first part of the tuple you'd have a parser that reads that initial byte and then maps it to the corresponding enum value or returns an error